### PR TITLE
[SPARK-8374] [YARN] Job frequently hangs after YARN preemption

### PR DESCRIPTION
**Pull request by xuchenCN**
Issue description [SPARK-8374](https://issues.apache.org/jira/browse/SPARK-8374)

Application starve because YARN scheduler preemption.
This PR fix two issues 
1. Application starve because CoarseGrainedSchedulerBackend.removeExecutor was happened before HeartbeatReceiver.killExecutorThread  

2.SpakrContext.killExecutor less executors than before BUT HeartbeatReceiver just want to expire the executor and expect new one

### CODE REVIEW

**Comment by xuchenCN** [comment]
CC @sryza

**Comment by andrewor14** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala [hunk 1]
```diff
@@ -47,4 +47,10 @@ private[spark] trait ExecutorAllocationClient {
    * @return whether the request is acknowledged by the cluster manager.
    */
   def killExecutor(executorId: String): Boolean = killExecutors(Seq(executorId))
+
+  /**
+   * Request that the cluster manager kill the specified executor and expect new one
+   * @return whether the request is acknowledged by the cluster manager.
+   */
+  def expireExecutor(executorId: String): Boolean
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala]: This is a confusing name. Without reading the java docs I have no idea what the difference between `killExecutor` and `expireExecutor` is

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -1439,6 +1439,24 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   @DeveloperApi
   override def killExecutor(executorId: String): Boolean = super.killExecutor(executorId)
 
+  /**
+   * :: DeveloperApi ::
+   * Request that the cluster manager kill the specified executor and expect new one
+   * @return whether the request is acknowledged by the cluster manager.
+   */
+  @DeveloperApi
+  override def expireExecutor(id: String): Boolean = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: We don't expect users to call this so we shouldn't expose this as a public API.

**Comment by andrewor14** [comment]
@xuchenCN The issue this patch is targeted for is a duplicate of SPARK-8119. There is already a fix for that in #7107. I believe the changes there are simpler and we should move forward with that one instead. Would you mind closing this PR?

**Comment by xuchenCN** [comment]
@andrewor14 All right close this PR

**Comment by tgravescs** [comment]
@xuchenCN can you please close this then

**Pull request closed without merging**